shared/js/event_safety.js

```diff
@@ -2,32 +2,57 @@
 /* exported eventSafety */
 
 /**
- * Guarantees a callback will be fired with a backing setTimeout.
+ * Listens for an event to fire _once_ or a timeout to expire.
  * This is generally used for things which are waiting on transitions
  * as it's sometimes possible to miss the transitionend event.
  * This should not be necessary once we have web animations as they
  * will guarantee that a callback is fired.
- * @param {Object} obj Object to listen to the event on.
- * @param {String} event The name of the event.
- * @param {Function} callback What to call after the event or timeout.
+ *
+ * Usage:
+ *
+ *   eventSafety(window, 'transitionend', callback, 400);
+ *
+ *   eventSafety(window, 'transitionend', 400).then(() => { ... });
+ *
+ * @param {Object} obj
+ *   Object to listen to the event on.
+ * @param {String} event
+ *   The name of the event.
+ * @param {Function} callback
+     Optional; called after the event is fired.
  * @param {Integer} timeout The duration of the safety timeout.
+ *
+ * @return {Promise}
+ *   A promise that will never reject.
  */
 function eventSafety(obj, event, callback, timeout) {
-  var finishTimeout;
-
-  function done(e) {
-
-    // transitionend events bubble by default, so we filter them by element.
-    if (e && e.type === 'transitionend' && e.target !== obj) {
-      return;
-    }
+  if (typeof callback === 'number') {
+    timeout = callback;
+    callback = null;
+  }
 
-    clearTimeout(finishTimeout);
-    obj.removeEventListener(event, done);
-    /*jshint validthis:true */
-    callback.apply(this, arguments);
+  if (!timeout) {
+    throw new Error('You must pass a valid timeout value to `eventSafety`.');
   }
 
-  obj.addEventListener(event, done);
-  finishTimeout = setTimeout(done, timeout);
+  return new Promise((resolve) => {
+    var finishTimeout;
+    function done(e) {
+      // transitionend events bubble by default, so we filter them by element.
+      if (e && e.type === 'transitionend' && e.target !== obj) {
+        return;
+      }
+
+      clearTimeout(finishTimeout);
+      obj.removeEventListener(event, done);
+      if (callback) {
+        /*jshint validthis:true */
+        resolve(callback.apply(this, arguments));
+      } else {
+        resolve(e);
+      }
+    }
+    finishTimeout = setTimeout(done, timeout);
+    obj.addEventListener(event, done);
+  });
 }
```

Comment:
please add unit test coverage for the returned promise (the lib tests are in `apps/sharedtest/test/unit/event_safety_test.js`)